pkg/referrer/manager.go

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (c) 2023. Nydus Developers. All rights reserved.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package referrer
+
+import (
+	"context"
+
+	"github.com/containerd/containerd/log"
+	"github.com/containerd/nydus-snapshotter/pkg/auth"
+	"github.com/golang/groupcache/lru"
+	"github.com/opencontainers/go-digest"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
+	"golang.org/x/sync/singleflight"
+)
+
+type Manager struct {
+	insecure bool
+	cache    *lru.Cache
+	sg       singleflight.Group
+}
+
+func NewManager(insecure bool) *Manager {
+	manager := Manager{
+		insecure: insecure,
+		cache:    lru.New(500),
+		sg:       singleflight.Group{},
+	}
+
+	return &manager
+}
+
+// CheckReferrer attempts to fetch the referrers and parse out
+// the nydus image by specified manifest digest.
+func (manager *Manager) CheckReferrer(ctx context.Context, ref string, manifestDigest digest.Digest) (*ocispec.Descriptor, error) {
+	metaLayer, err, _ := manager.sg.Do(manifestDigest.String(), func() (interface{}, error) {
+		// Try to get nydus metadata layer descriptor from LRU cache.
+		if metaLayer, ok := manager.cache.Get(manifestDigest); ok {
+			desc := metaLayer.(ocispec.Descriptor)
```

Comment:
Should check if type assertion is failed?

Reply:
No need, it must be `ocispec.Descriptor` if `ok == true`.